source/agora/consensus/protocol/Nominator.d

```diff
@@ -673,28 +674,34 @@ extern(D):
     public override Value combineCandidates (uint64_t slot_idx,
         ref const(set!Value) candidates)
     {
-        scope (failure) assert(0);
-
-        foreach (ref const(Value) candidate; candidates)
+        try
         {
-            auto data = deserializeFull!ConsensusData(candidate[]);
-
-            if (auto msg = this.ledger.validateConsensusData(data))
+            foreach (ref const(Value) candidate; candidates)
             {
-                log.error("combineCandidates(): Invalid consensus data: {}", msg);
-                continue;
+                auto data = deserializeFull!ConsensusData(candidate[]);
+
+                if (auto msg = this.ledger.validateConsensusData(data))
+                {
+                    log.error("combineCandidates(): Invalid consensus data: {}", msg);
+                    continue;
+                }
+                else
+                {
+                    log.info("combineCandidates: {}", slot_idx);
+                }
+
+                // todo: currently we just pick the first of the candidate values,
+                // but we should ideally pick tx's out of the combined set
+                return duplicate_value(&candidate);
             }
-            else
-            {
-                log.info("combineCandidates: {}", slot_idx);
-            }
-
-            // todo: currently we just pick the first of the candidate values,
-            // but we should ideally pick tx's out of the combined set
-            return duplicate_value(&candidate);
+        } catch (Exception ex)
+        {
+            assert(0, format!"[%s]:[%s] combineCandidates: slot %u. Exception:%s"(
+                __FILE__, __LINE__, slot_idx, ex.to!string));
         }
-
-        assert(0);  // should not reach here
+        // should not reach here
+        assert(0, format!"[%s]:[%s] combineCandidates: no valid candidate for slot %u."(
+            __FILE__, __LINE__, slot_idx));
```

Comment:
I think we don't need the file name and line for logging because the logger already shows the information.

Reply:
For these assertion messages we want to be able to click on the file and line number to jump to the code in the IDE.

Reply:
What he means is that the following code:
```D
void main ()
{
 	assert(0);   
}
```

Will give you this by default (try it on run.dlang.org):
```
core.exception.AssertError@onlineapp.d(3): Assertion failure
----------------
??:? _d_assertp [0x56302dc82be1]
onlineapp.d:3 _Dmain [0x56302dc82b10]
```

You can see that even though no debug info is present, the file/line info is (`onlineapp.d(3)`).

Reply:
Yes but try clicking on that in an IDE and it will not jump to the line of code.

Reply:
Well then the IDE should be fixed, not the codebase itself. For example, to make jumping to file & line of errors work for me in Sublime I have:

> "file_regex": "^([^\\(]+)\\((\\d+,\\d+)\\): (.*)$",

I didn't go into the codebase and adapted it to serve my own purpose.

Reply:
Fair point. I will make an update before the end of the sprint to remove these.